Compound order for histogram aggregations

qwerty4030 opened the pull request:
This PR adds support for `histogram` and `date_histogram` agg compound `order` by refactoring and reusing `terms` agg `order` code. The major change is that the `Terms.Order` and `Histogram.Order` classes have been replaced/refactored into a new class `BucketOrder`. This is a **breaking change** for the Java Transport API. For backward compatibility with previous ES versions the `(date)histogram` compound `order` will use the first order. Also the `_term` and `_time` aggregation order keys have been **deprecated**; replaced by `_key`.

Relates to #20003: now that all these aggregations use the same `order` code, it should be easier to move validation to parse time (as a follow up PR).

Relates to #14771: `histogram` and `date_histogram` aggregation `order` will now be validated at reduce time.

Closes #23613: if a single `BucketOrder` that is not a tie-breaker is added with the Java Transport API, it will be converted into a `CompoundOrder` with a tie-breaker.

qwerty4030 commented:
@colings86 would you mind taking a quick look? I need to rebase on master due to major changes in the parsers, but I want to make sure I'm heading in the right direction :) Thanks!

qwerty4030 commented:
@colings86 any chance you would be able to take a look this week? Thanks!

colings86 commented:
@qwerty4030 Sorry I have been a bit slow getting to review this. I will try to get to it either later this week or next week.

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/Histogram.java:
```diff
@@ -103,13 +107,27 @@ public int compare(Histogram.Bucket b1, Histogram.Bucket b2) {
             }
         });
 
+        public static final OrderSpec SPEC = new OrderSpec() {
```
I'm not sure I follow what purpose this servers, could you explain?

qwerty4030 replied:
I created this interface to help centralize the code that depends on aggregation specific orders. e.g. `histogram` aggs support a `_key` or `_time` order and `terms` aggs support a `_term` order. The serialization logic (`byte id` values) was different and the `CompoundOrder` sometimes needs to automatically add a "tie-breaker to avoid non-deterministic ordering if all user provided comparators return 0." Which is `_key` ascending for `histogram` aggs and `_term` ascending for `terms` aggs.

`OrderSpec#getOrders()` returns all of the aggregation specific sorts so `InternalOrder.Streams#readOrder` can deserialize a `byte id` back into the correct `BucketOrder`. Note aggregation and compound order serialization was the same for `terms` and `histogram` so I moved that code into that `readOrder` method.

`OrderSpec#getAbsoluteOrder` returns the "tie-breaker" `BucketOrder` needed in `CompoundOrder`

Each agg creates its own `OrderSpec` instance that is required when creating a `CompoundOrder` or deserializing a `BucketOrder`.

Some thoughts:
I can change `OrderSpec#getOrders()` into something like `OrderSpec#getOrder(byte id)` if that makes more sense.
Another solution may be to serialize the `key` and `asc` values instead of just an ID.

colings86 replied:
I wonder if it would be better to move towards supporting only `_key` for the time/term ordering since to the use this is actually what its doing, sorting by the key of the bucket? We could do this in a backward compatible way by using the`ParseField` and having `_time` and `_key` as deprecated values for the order. We could also solve the mismatch in `byte id` values by checking the version in the writeTo/readFrom methods and if the version is <6.0 making the code read/write the old id values and otherwise read/write a consistent response. This should mean that we can move towards only having BucketOrder and not the individual implementations.

qwerty4030 replied:
I see. So there will only be 4 types of orders supported:
1. `_key`: order by bucket key (`_term` and `_time` would be a deprecated alias to this)
2. `_count`: order by bucket doc count
3. order by aggregation (assume this sort if order key does not match any of the above reserved sort keys)
4. compound: sort by a list of any combination of the sorts above (`order` param is an array)

If it is safe to assume this model will fit the current and future agg order features, then I can move much more the code into the common `BucketOrder` class. I may need to fight with the generics a bit (used for the `Comparator`) but I don't think its too bad.

Good idea with serialization logic checking the version. This way I can standardize the IDs for the 4 sorts above and remove that `OrderSpec` interface. Thanks for the input.

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java:
```diff
@@ -202,9 +208,9 @@ public InternalAggregation doReduce(List<InternalAggregation> aggregations, Redu
+            if (terms.getBucketsInternal().size() < getShardSize() || Terms.Order.isTermOrder(order)) {
                 thisAggDocCountError = 0;
-            } else if (InternalOrder.isCountDesc(this.order)) {
+            } else if (Terms.Order.isCountDesc(this.order)) {
```
Could you explain why `BucketOrder` can't be used here?

qwerty4030 replied:
The `_count` order is specific to the `terms` agg (code treats it like the `_term` order). Also this method compares it directly to `Terms.Order#COUNT_DESC` instance so thats why this code references `Terms.Order` instead of something in `BucketOrder`.

It might be possible to move the `_count` order code into a common place like `BucketOrder` but that would require more changes (looks like `terms` and `histogram` aggs use a different `byte id` for `_count` order serialization.)

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/OrderSpec.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.List;
+
+public interface OrderSpec {
```
Could you add JavaDocsfor this interface and its methods?

qwerty4030 replied:
Definitely will do. Sorry I should have added it sooner; made the review harder than it should have been! See my comment [here](https://github.com/elastic/elasticsearch/pull/22343#discussion_r97937099) for more info in the mean time.

colings86 commented:
To Answer your specific questions above:

> 1. Parser code is difficult to reuse due to the complex if else chains/trees.
Agreed, there is an `ObjectParser` class now that makes building parsers much easier, maybe we could/should use that here? I think most of the aggregation code has been moved over to use this now anyway

> 3. I named the class BucketOrder instead of Order to avoid refactoring all the references to the static methods used in builders, e.g. terms(....).order(Terms.Order.term(true))). Otherwise I would have to use the full package name in the code.
Sounds good. I think that `BucketOrder` makes more sense anyway

> 4. I'm pretty sure this is a breaking change because the serialization changes for the histogram aggs and the return type changes for those static builder methods mentioned above.
Yes, I think this will probably be a breaking change as is. We will need to look into backwards compatibility later before we merge.

> 5.Not sure why but the TermsParser puts this intermediate List<OrderElement> in the otherOptions instead of just directly using the CompoundOrder?
I'm not sure why it does this either, maybe its something we can avoid now and just use `CompoundOrder` directly?

> 6. A special internal optimization only for the terms agg is in InternalOrder.Aggregation#comparator. This is not possible for the histogram aggs because the buckets do not have a bucketOrd. Should I look into adding this?
I think this optimisation is only important for the terms aggregation because of the `breadth_first` mode (which is only available for the terms aggregation) but maybe @jpountz can confirm/give more details?

jpountz commented:
> I think this optimisation is only important for the terms aggregation because of the breadth_first mode (which is only available for the terms aggregation) but maybe @jpountz can confirm/give more details?

This optimization is only useful to sort at the shard level. This is needed for terms since not all terms need to be returned to the coordinating node so we must select the best ones at the shard level. However histograms are much simpler since all buckets are always returned, so we can just do the sorting on the coordinating node and that optimization is irrelevant.

qwerty4030 commented:
@colings86 @jpountz Thanks for the input guys! FYI I'm going to be away from my computer for a couple weeks, then I'll get back to this. There seems to be quite a few changes in master in the mean time. Do you recommend I rebase on latest master before updating this PR? Thanks again.

colings86 commented:
@qwerty4030 yeah I thin it would be best to rebase your branch and go form there

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -295,30 +234,106 @@ public boolean equals(Object obj) {
+
+    private static Comparator comparingKeys() {
+        return (o1, o2) -> {
+            if (o1 instanceof Terms.Bucket && o2 instanceof Terms.Bucket) {
```
any ideas how to make this `Comparator` cleaner to avoid using `instanceof`?

colings86 replied:
I don't we'll find it easy to get rid of the `instanceof`'s in this PR but maybe in the future we can make the buckets `Comparable` and then we can just use the `compareTo` method?

qwerty4030 replied:
Not a bad idea. I'm assuming the `compareTo` method would only compare the bucket keys?

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -295,30 +234,106 @@ public boolean equals(Object obj) {
+        };
+    }
+
+    public static boolean isCountDesc(BucketOrder order) {
```
These two methods `isCountDesc` and `isKeyOrder` are checking for some very specific classes and seem to missing some general ones, e.g. they will return false if passed a `Compound` order will the wrong number of elements (even if it meets the "definintion" of the method name (I couldn't find any other comments). These only seem to be used from the `terms` agg code,

colings86 replied:
Not sure what you mean here? Do you mean it will return false if its passed a `Compound` order with only 1 element which is `InternalOrder.COUNT_DESC`?

qwerty4030 replied:
Exactly, or a `CompoundOrder` with `InternalOrder.COUNT_DESC` and then an `Aggregation` Order. The reason I brought it up was I wanted to simply these two methods to just check if the first order is the one its looking for.

qwerty4030 replied:
I updated these methods in b1bb5a9fef9e35846ce046885401ed9e2527dc8e.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -327,40 +342,82 @@ public static void writeOrder(Terms.Order order, StreamOutput out) throws IOExce
-
-        public static Terms.Order readOrder(StreamInput in, boolean absoluteOrder) throws IOException {
+        public static BucketOrder readOrder(StreamInput in) throws IOException {
+            // TODO backwards compat
```
any examples of the proper way to do backwards compatibility code?

colings86 replied:
You'll need to add a check for the version from the stream, so something like:
```
if (in.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
    <NEW SERIALISATION CODE GOES HERE>
} else {
    <BACKWARDS COMPATIBLE SERIALISATION CODE GOES HERE>
}
```

Have a look at the `IncludeExclude` class for and example. You will obviously need to do a similar thing for the `writeTo` method

qwerty4030 replied:
What is the expected behavior for the `writeTo` method for the `histogram` if the `order` is a `CompundOrder`? I think the best option is just to serialize the first order? However wont this produce unexpected results if the cluster has mixed versions? I'm not sure how the transport client works in this case.

qwerty4030 replied:
@colings86 made the changes here: https://github.com/elastic/elasticsearch/pull/22343/commits/fa54c8e037017bb9826082e59d0f741bfa300ebe
See this line for the `writeTo` backward compat logic that I was not sure about: https://github.com/elastic/elasticsearch/pull/22343/commits/fa54c8e037017bb9826082e59d0f741bfa300ebe#diff-78913b4f2494b84e11cc7cc3e12c5648R349
Still need to update documentation for the breaking Java API changes.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -327,40 +342,82 @@ public static void writeOrder(Terms.Order order, StreamOutput out) throws IOExce
+
+        static BucketOrder resolveOrder(String key, boolean asc) {
+            switch(key) {
+                case "_term": // TODO deprecate
```
any examples of the proper way to do deprecation warnings and logging?

colings86 replied:
Deprecation logging needs to be done using the `DeprecationLogger` class. If you look at `PutIndexTemplateRequest` you'll see an example. That should also add a warning header to the response.

Other things we need to do for deprecation include adding a deprecation note in the relevant place int he documentation, and adding the description of any breaking changes to the breaking changes document

qwerty4030 replied:
See changes here: https://github.com/elastic/elasticsearch/pull/22343/commits/0fc36a9b2581a089e833709fa763525c4499a114
Still need to add deprecation note to documentation.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/BucketOrder.java:
```diff
@@ -0,0 +1,107 @@
+import java.util.Comparator;
+import java.util.List;
+
+public abstract class BucketOrder<B extends Bucket> implements ToXContent {
```
technically I can changes this class name back to `Order` since I removed the other agg specific `Order` classes. or is `BucketOrder` a better name?

qwerty4030 commented:
@colings86 Would you mind taking another look? I completely removed the agg specific `Order` classes, so now pretty much all code uses `BucketOrder`. A bunch of test code had to be updated, but its not too bad (just renaming classes/methods). Code is still WIP, here are the TODOs:
- backwards compat code
- deprecate _term and _time order (now just _key)
- code cleanup and java doc
- update documentation/release notes
- tests

If I could get the answers to those questions I left on the commit that would be great. Thank you!

colings86 commented:
@qwerty4030 Thanks for making the updates. I'm going to be away for a couple of weeks now but I will make sure I take a look when I get back.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/InternalTerms.java:
```diff
@@ -112,6 +113,10 @@ public long getDocCountError() {
             return docCountError;
         }
 
+        public long getBucketOrd() {
```
Its used in the special case code just for the `terms` agg in `InternalOrder.Aggregation#comparator`.
I can move that code to some static method in this `terms` package, so it can access `InternalTerms.Bucket#bucketOrd` like it did before.
Would `TermsAggregator#comparator` be a good place for it?

qwerty4030 replied:
Removed this method and refactored the code here: https://github.com/elastic/elasticsearch/pull/22343/commits/8ddaf23f9d181015e6e38941c786c78500a5159c

qwerty4030 commented:
@colings86 I noticed you removed the `breaking` label. I will add backwards compatibility code to support previous ES versions in the cluster, but the Java transport API has changed so some users will get compile errors if they used the `terms` and `histogram` `Order` classes (the ones I had to update in the tests). Is that still considered a breaking change?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

colings86 commented:
@qwerty4030 I removed the `breaking` label because the PR will add backwards compatibility so the wire protocol and REST API don't break. We use a separate label for changes that only break the Java API and I have added it to this PR as you rightly point out that the Java API will change with this PR.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -0,0 +1,603 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
What is the purpose of the `hashCode()` and `equals` methods here? How should they be tested?

colings86 replied:
These are used in unit tests so we can check that when objects are serialised and then deserialised the resulting object is equal to the original

qwerty4030 replied:
I'm assuming this was never tested? The original code did not check the `key`. This meant that the `Aggreggation` orders were always equal, even with different paths. I will add tests for this in the new `InternalOrderTests`,

colings86 replied:
Yeah it looks like there wasn't a unit test for this before, so I'm +1 on adding an `InternalOrderTests`.

qwerty4030 replied:
done in 024f343

qwerty4030 commented on docs/reference/aggregations/bucket/histogram-aggregation.asciidoc:
```diff
@@ -179,120 +179,8 @@ POST /sales/_search?size=0
-
-The above will sort the buckets based on the avg rating among the promoted products
-
+By default the returned buckets are sorted by their `key` ascending, though the order behaviour can be controlled using
```
Not sure if its better to just reference `terms` agg `order` docs or to create a new separate file that is included in all 3 aggregations?

colings86 replied:
hmmm I wonder if it might even be better to just duplicate the explanation of the order syntax here to make it easier for users to read? @jpountz what do you think

qwerty4030 replied:
@colings86 @jpountz Should I just duplicate the docs for all the aggs that support `order`? Is it possible to include a shared file in here?

qwerty4030 commented:
@colings86 Ready for a full review. All code, docs, and tests have been updated. I left a couple more questions, but everything else looks good to me. Thanks!

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -0,0 +1,603 @@
+         * @throws IOException on error reading from the stream.
+         */
+        public static BucketOrder readHistogramOrder(StreamInput in) throws IOException {
+            if (in.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
```
How/where should the backwards compat code be tested?

colings86 replied:
I think you should be able to do this in a unit test for this class when you create the `StreamInput` you should be able to set the version

qwerty4030 replied:
done in 024f343

qwerty4030 commented:
@colings86 Any chance for a review soon? Conflicts are pilling up again. Thanks!

colings86 commented:
@qwerty4030 sure, its on my list to get done tomorrow

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -0,0 +1,603 @@
+/**
+ * Implementations for {@link Bucket} ordering strategies.
+ */
+public class InternalOrder extends BucketOrder {
```
Could you add a unit test for this new class named `InternalOrderTests`? I think it should extend `AbstractSerialisingTestCase` so we can use it to test the XContent and Writeable serialisation

qwerty4030 replied:
@colings86 I ran into a few roadblocks trying to use `AbstractSerialisingTestCase`:

1. It requires the type to implement `Writeable`. However this is not too hard to add, unless it needs that backwards compat logic.
2. The test checks that the object that was serialized is **not** the same object that was deserialized. However the `_count` and `_key` are static `BucketOrder` instance so the test fails. I don't see an easy way of getting around this.
3. I think `BucketOrder` should implement `ToXContentObject` instead of just `ToXContent`. It outputs an array for the `CompoundOrder` and full objects for the other `InternalOrder`s. I'm not sure of the side effects of changing this, but without it the `AbstractSerialisingTestCase` fails due to not expecting an array/object start token.

qwerty4030 replied:
4. Looks like there is bug in `ESTestCase#shuffleXContent()` which is called from `AbstractSerialisingTestCase`. It does not support `XContent` as lists/arrays, e.g. `CompoundOrder`, and the test fails because this method shuffles an empty map. I managed to hack in support for a list, but noticed there were more bugs: e.g. `ESTestCase#shuffleMap` doesnt check if the value is a `List` (which may contain a `Map` that should be shuffled?)

qwerty4030 replied:
@colings86 I resolved 1,3, and 4 in 024f343. But I still have the issue with the test checking that the objects must not be equal. Any ideas?

colings86 replied:
@qwerty4030 This is indeed an issue. I spoke to @javanna about this and the only thing we could come up with is to make `AbstractSerialisingTestCase.assertParsedInstance()` a `protected` method instead of `private` and override it in your `InternalOrderTests` to deserialise the bytes into a new instance but not do the `assertNotSame` check.

qwerty4030 replied:
done in 4c5fec4

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -0,0 +1,603 @@
+            if (b1 instanceof Terms.Bucket) {
+                return ((Terms.Bucket) b1).compareTerm((Terms.Bucket) b2);
+            } else if (b1 instanceof InternalHistogram.Bucket) {
+                return Double.compare(((InternalHistogram.Bucket) b1).key, ((InternalHistogram.Bucket) b2).key);
```
Yup that would work, however this will make this code pretty ugly with even more casts. I'm starting to really like your original suggestion of making the buckets `implements Comparable`.

colings86 replied:
> I'm starting to really like your original suggestion of making the buckets implements Comparable.

I think it might be worth exploring, hopefully it will make the code a bit cleaner

qwerty4030 replied:
done in 4c5fec4
I created a new interface `KeyComparable` for this. IMO the code is much cleaner now (less casts and that `key` field is no longer `public`.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java:
```diff
@@ -0,0 +1,603 @@
+            } else if (b1 instanceof InternalHistogram.Bucket) {
+                return Double.compare(((InternalHistogram.Bucket) b1).key, ((InternalHistogram.Bucket) b2).key);
+            } else if (b1 instanceof InternalDateHistogram.Bucket) {
+                return Long.compare(((InternalDateHistogram.Bucket) b1).key, ((InternalDateHistogram.Bucket) b2).key);
```
done in 4c5fec4 see my comment above

qwerty4030 commented on core/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java:
```diff
@@ -107,6 +110,12 @@ protected FunctionScoreQueryBuilder doCreateTestQueryBuilder() {
     @Override
+    protected String[] shuffleProtectedFields() {
+        // do not shuffle fields that may contain arbitrary content
+        return SHUFFLE_PROTECTED_FIELDS;
```
Had to make this change now that `XContent` is shuffled for objects inside lists.

qwerty4030 commented on core/src/test/java/org/elasticsearch/index/query/MoreLikeThisQueryBuilderTests.java:
```diff
@@ -204,6 +207,16 @@ protected MoreLikeThisQueryBuilder doCreateTestQueryBuilder() {
+     */
+    @Override
+    protected String[] shuffleProtectedFields() {
+        return SHUFFLE_PROTECTED_FIELDS;
```
Had to make this change now that `XContent` is shuffled for objects inside lists.

qwerty4030 commented:
@colings86 Addressed all comments except the one about the duplicate documentation. Still waiting on @jpountz to take a look at that was well as the rest of this PR?

colings86 commented:
Jenkins test this

colings86 commented:
jenkins retest this

qwerty4030 commented:
@colings86 What are the next steps? I noticed some test fails in that jenkins build, but this branch is pretty old and I use an older gradle 2.13 on my local machine. Let me know if I should squash+rebase on master or just merge master in. Thanks.

colings86 commented:
@qwerty4030 it's probably worth rebasing/merging (which ever you prefer) this branch with the latest master and then running `gradle check` to see whether the build failures are due to this change and fixing if necessary. I would use gradle 3.3 for this.

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:
```diff
@@ -244,17 +246,32 @@ public DateHistogramAggregationBuilder extendedBounds(ExtendedBounds extendedBou
+        if (orders == null) {
+            throw new IllegalArgumentException("[orders] must not be null: [" + name + "]");
+        }
+        order(orders.size() > 1 ? BucketOrder.compound(orders): orders.get(0));
```
should we just expect users to do this by themselves?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregationBuilder.java:
```diff
@@ -185,17 +186,32 @@ public HistogramAggregationBuilder extendedBounds(double minBound, double maxBou
+        if(order instanceof CompoundOrder || InternalOrder.isKeyOrder(order)) {
+            this.order = order;
+        } else {
+            this.order = BucketOrder.compound(order);
```
can you ad a comment about why we do this?

qwerty4030 replied:
Will do. What are your thoughts about refactoring these `order()` methods to be reusable? The logic repeated in all 3 aggs that support `order`.

jpountz replied:
actually I don't fully understand why we can't just do `this.order = order` all the time

qwerty4030 replied:
We may need to add a tie-breaker sort (`_key` asc) to prevent non-deterministic ordering. See #23613

qwerty4030 replied:
done in 398e279

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramAggregationBuilder.java:
```diff
@@ -185,17 +186,32 @@ public HistogramAggregationBuilder extendedBounds(double minBound, double maxBou
+        if (orders == null) {
+            throw new IllegalArgumentException("[orders] must not be null: [" + name + "]");
+        }
+        order(orders.size() > 1 ? BucketOrder.compound(orders) : orders.get(0));
```
let's leave this responsibility to users?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregationBuilder.java:
```diff
@@ -192,29 +192,33 @@ public TermsAggregationBuilder shardMinDocCount(long shardMinDocCount) {
+            this.order = order;
+        } else {
+            this.order = BucketOrder.compound(order);
+        }
```
please add a comment to explain

qwerty4030 replied:
done in 398e279

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregationBuilder.java:
```diff
@@ -192,29 +192,33 @@ public TermsAggregationBuilder shardMinDocCount(long shardMinDocCount) {
             throw new IllegalArgumentException("[orders] must not be null: [" + name + "]");
         }
-        order(Terms.Order.compound(orders));
+        order(orders.size() > 1 ? BucketOrder.compound(orders) : orders.get(0));
```
oh now I understand why you were adding this method in other aggs as well. @colings86 what do you think, should we keep it or remove it?

qwerty4030 replied:
This method has to handle a few different things

1. Handle a list of orders from parsed from xcontent (list or single order)
2. Add tie breaker `_key` order if needed
3. consistently convert the agg builder to/from xcontent

The old code had a bug with 3 if `order(BucketOrder order)` was called instead of `order(List<BucketOrder> orders)` because the later was used to parse the xcontent which resulted in a compound order instead of the individual order. This is now fixed and the test now calls both of these `order()` methods (randomly).

qwerty4030 replied:
Also I am not against making this method `private` (only called by parser). That way only the one `order(BucketOrder order)` method is exposed to the user to avoid confusion between `List<BucketOrder>` and `BucketOrder.compound(BucketOrder...)`

qwerty4030 replied:
@jpountz @colings86 Thoughts on this `order(List<BucketOrder> orders)` method? Would we prefer to have it `private` or in some common code shared by the aggs that use `order`?

jpountz replied:
I think it's fine this way. This PR is already good incremental progress.

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/support/BucketOrder.java:
```diff
@@ -0,0 +1,126 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.support;
```
This package has only been used for internal classes for now. I'm wondering whether it should be directly under o.e.s.aggs

qwerty4030 replied:
Sure why not. I'll also move the `interface KeyComparable` to the `org.elasticsearch.search.aggregations` package as well?

qwerty4030 replied:
@jpountz I forgot that this class references some package-private fields/methods in `InternalOrder`. Obviously I can make all that public or move `InternalOrder` as well or do you prefer to keep the package as is?

jpountz replied:
I'd move InternalOrder as well.

qwerty4030 replied:
done in 61ace2f

qwerty4030 commented:
I merged in the latest master branch and ran `gradle check` successfully.

qwerty4030 commented:
Had to merge master again to fix BWC tests. I'm not sure why they started failing, but `gradle check` is successful again.

qwerty4030 commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateHistogramAggregationBuilder.java:
```diff
@@ -137,9 +139,7 @@ public DateHistogramAggregationBuilder(String name) {
     /** Read from a stream, for internal use only. */
     public DateHistogramAggregationBuilder(StreamInput in) throws IOException {
         super(in, ValuesSourceType.NUMERIC, ValueType.DATE);
-        if (in.readBoolean()) {
```
I decided to remove this boolean flag in the (date) histogram agg builders since it is always true (we always have an order). This made the BWC code a bit more complicated (for some reason `InternalHistogram` did not use this flag), but simplified the new read/write logic.

colings86 commented:
jenkins retest this

qwerty4030 commented:
@colings86 build looks good now? I left a couple questions on some unresolved comments. Thanks!

qwerty4030 commented:
Merged master again to resolve conflicts and ran `gradle check` successfully.

colings86 commented:
@qwerty4030 I think everything looks good with this PR now, thanks for your patience getting it reviewed and sorted. Would you be able to merge master one more time and I'll make sure I kick of a build and then try to merge this in quickly before any more conflicts appear?

qwerty4030 commented:
@colings86 Sounds good. Is it alright if I squash the commits and rebase on master instead? @jpountz Thanks for reviewing as well.

colings86 commented:
>  Is it alright if I squash the commits and rebase on master instead?

Yes thats fine it if you find that easier

qwerty4030 commented:
@colings86 Rebase done.

colings86 commented:
jenkins retest this

colings86 commented:
@qwerty4030 thanks, I've kicked off a build and will merge this PR once it passes

colings86 commented:
@qwerty4030 merged, thanks for your contribution 😄

nik9000 commented:
I wonder if we can make a change to https://raw.githubusercontent.com/elastic/elasticsearch/master/docs/reference/migration/migrate_6_0/java.asciidoc to help folks use this.

colings86 commented:
@nik9000 this PR did add a paragraph in https://github.com/elastic/elasticsearch/pull/22343/files#diff-ae29a8e58e261b8425a707923a4e71bdR33 . Is there something in particular you think should be made clearer?

nik9000 commented:
Nope! Sorry, I was searching in the PR and didn't see them at first. I should have looked more closely.

qwerty4030 commented:
@nik9000 I also updated the java api docs here: https://www.elastic.co/guide/en/elasticsearch/client/java-api/master/_bucket_aggregations.html#_order
But you did get me thinking, how will the other official clients get updated (java REST, js, etc...)? Should we create issues for each one? How does this usually work with new/deprecated features?
